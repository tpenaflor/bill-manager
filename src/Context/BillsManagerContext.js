import React from 'react'

const BillsManagerContext = React.createContext()

export default BillsManagerContext