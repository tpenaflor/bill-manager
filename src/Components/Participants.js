import React from 'react'
import {useBillsManagerContext} from '../Context/BillsManagerContext'

const Participants = () => {

  const {bills, participants, partDispatch, billDispatch} = useBillsManagerContext()
  const remParticipant = (id) => {
    partDispatch({
      type:"DEL_PARTICIPANT",
      id
    })
    
    bills.forEach((bill)=> {
      let unitCount
      if (bill.splitOpt === 'even') {
        unitCount = bill.allocation.participant.length - 1
      }
      else {
        unitCount = bill.allocation.unitCount
      }

      billDispatch(
        {
          type : "DEL_PARTICIPANT_FROM_BILL",
          id : bill.id,
          participant : {
            id
          },
          unitCount
        }
      )
    })
  }

  return (
    <div className="entry">
      <div> </div>
      {participants.map((part)=> (
        <div key={part.id}>
          <button onClick={()=>remParticipant(part.id)} >x</button>
          <p>{part.name}</p>
        </div>
      ))}
    </div>
  )
}

export default Participants