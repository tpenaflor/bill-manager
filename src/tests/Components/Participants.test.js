import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json';

import Participants from '../../Components/Participants'
import * as BillsManagerContext from '../../Context/BillsManagerContext'
import { billFixture} from '../Fixtures/BillFixture'
import {participantFixture} from '../Fixtures/ParticipantFixture'

let wrapper, partDispatch, billDispatch

beforeEach(() => {
  partDispatch = jest.fn()
  billDispatch = jest.fn()

  const contextValues = {
    partDispatch,
    billDispatch,
    bills: billFixture,
    participants: participantFixture
  }

  jest.spyOn(BillsManagerContext, 'useBillsManagerContext')
    .mockImplementation(() => contextValues)

  wrapper = shallow( <Participants / > )
  expect(shallowToJson(wrapper)).toMatchSnapshot()
})

afterEach(() => {
  expect(shallowToJson(wrapper)).toMatchSnapshot()
})

test('remove participant', () => {
  wrapper.find('button').at(1).simulate('click')

  expect(partDispatch).toHaveBeenLastCalledWith({
    type: "DEL_PARTICIPANT",
    id : participantFixture[1].id
  })

  expect(billDispatch).toHaveBeenCalledWith({
    type : "DEL_PARTICIPANT_FROM_BILL",
    id: billFixture[1].id,
    participant: {
      id : participantFixture[1].id
    },
    unitCount : billFixture[1].allocation.unitCount
  })

  expect(billDispatch).toHaveBeenCalledWith({
    type : "DEL_PARTICIPANT_FROM_BILL",
    id: billFixture[3].id,
    participant: {
      id : participantFixture[1].id
    },
    unitCount : billFixture[3].allocation.unitCount - 1
  })

})