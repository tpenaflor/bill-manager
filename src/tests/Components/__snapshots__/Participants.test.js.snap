// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remove participant 1`] = `
<div
  className="entry"
>
  <div>
     
  </div>
  <div
    key="a4714460-a11c-4681-a21d-c7de2c57a02c"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user1
    </p>
  </div>
  <div
    key="ac33c426-53c4-4d38-ba0f-f8e569a9a507"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user2
    </p>
  </div>
  <div
    key="3712fdf0-5b62-41b6-9442-b35a41d5ec81"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user3
    </p>
  </div>
  <div
    key="6be42a94-61eb-47d4-8552-0369487f015c"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user4
    </p>
  </div>
</div>
`;

exports[`remove participant 2`] = `
<div
  className="entry"
>
  <div>
     
  </div>
  <div
    key="a4714460-a11c-4681-a21d-c7de2c57a02c"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user1
    </p>
  </div>
  <div
    key="ac33c426-53c4-4d38-ba0f-f8e569a9a507"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user2
    </p>
  </div>
  <div
    key="3712fdf0-5b62-41b6-9442-b35a41d5ec81"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user3
    </p>
  </div>
  <div
    key="6be42a94-61eb-47d4-8552-0369487f015c"
  >
    <button
      onClick={[Function]}
    >
      x
    </button>
    <p>
      user4
    </p>
  </div>
</div>
`;
